trie/transition.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2021 go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package trie
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethdb"
+)
+
+type TransitionTrie struct {
+	overlay *VerkleTrie
+	base    *SecureTrie
+}
+
+func NewTransitionTree(base *SecureTrie, overlay *VerkleTrie) *TransitionTrie {
+	return &TransitionTrie{
+		overlay: overlay,
+		base:    base,
+	}
+}
+
+func (t *TransitionTrie) Base() *SecureTrie {
+	return t.base
+}
+
+// GetKey returns the sha3 preimage of a hashed key that was previously used
+// to store a value.
+//
+// TODO(fjl): remove this when StateTrie is removed
+func (t *TransitionTrie) GetKey(key []byte) []byte {
+	if t.overlay.GetKey(key) != nil {
+		return t.overlay.GetKey(key)
+	}
+	return t.base.GetKey(key)
+}
+
+// TryGet returns the value for key stored in the trie. The value bytes must
+// not be modified by the caller. If a node was not found in the database, a
+// trie.MissingNodeError is returned.
+func (t *TransitionTrie) TryGet(addr, key []byte) ([]byte, error) {
+	if t.overlay.GetKey(key) != nil {
+		return t.overlay.TryGet(addr, key)
+	}
+	// TODO also insert value into overlay
```

Comment:
Tangent thought: doing this vs not doing may change the gas fee dynamics for users. Not sure if that's relevant for deciding doing this or not (from my perspective it would be better doing it, as we talked). But... just sharing the thought if UX could be impacted in a negative way.

Reply:
well technically no gas will be charged. So if we change the gas costs during the transition (we should), then we should charge a blanket cost of : reading into vkt + reading into mpt